rfd/0202-db-multi-session-mfa.md

```diff
@@ -0,0 +1,309 @@
++   require_session_mfa: "multi-session"
+```
+
+Mode defaults to per-session if not set. If a resource matches a role set with
+per-session but others on `multi-session`, the stricter mode should be applied.
+
+The multi-session MFA extends [RFD 155 Scoped Webauthn
+Credentials](https://github.com/gravitational/teleport/blob/master/rfd/0155-scoped-webauthn-credentials.md)
+with a new scope for executing database sessions:
+```diff
+// webauthn.proto
+enum ChallengeScope {
+...
++    // Used for 'tsh db exec' and allows reuse. 
++    SCOPE_DATABASE_MULTI_SESSION = 8;
+}
+```
+
+Similar to `SCOPE_ADMIN_ACTION`, the new scope `SCOPE_DATABASE_MULTI_SESSION`
+will allow reuse of the MFA session data until it expires (5 minutes for
+WebAuthn). Clients must go through MFA ceremony again if it expires.
+
+The MFA response will be checked upon auth call of `GenerateUserCerts` where
+user requests a TLS user cert with database route. New logic is added to
+`GenerateUserCerts` where the new scope with reuse is allowed only if the role
+set matching the requested database has `roleset.option.require_session_mfa`
+option set to `multi-session`.
+
+The new scope cannot be used for `GenerateUserCerts` for non-database targets.
+And if the MFA response is validated with existing non-reusable `SCOPE_SESSION`,
+the action should always be allowed.
+
+Here is a quick matrix:
+
+| `require_session_mfa` mode | MFA response scope             | Requested Target | Access |
+|----------------------------|--------------------------------|------------------|--------|
+| `multi-session`            | `SCOPE_SESSION`                | Database         | Allow  |
+| `multi-session`            | `SCOPE_DATABASE_MULTI_SESSION` | Database         | Allow  |
+| `multi-session`            | `SCOPE_DATABASE_MULTI_SESSION` | Non-Database     | Denied |
+| `session`                  | `SCOPE_SESSION`                | Database         | Allow  |
+| `session`                  | `SCOPE_DATABASE_MULTI_SESSION` | Database         | Denied |
+| `session`                  | `SCOPE_DATABASE_MULTI_SESSION` | Non-Database     | Denied |
+
+MFA requirement check is also updated to indicate whether the client can get
+away with reusing the MFA response.
+```diff
+// MFARequired indicates if MFA is required to access a
+// resource.
+enum MFARequired {
+  // Indicates the client/server are either old and don't support
+  // checking if MFA is required during the ceremony or that there
+  // was a catastrophic error checking RBAC to determine if completing
+  // an MFA ceremony will grant access to a resource.
+  MFA_REQUIRED_UNSPECIFIED = 0;
+  // Completing an MFA ceremony will grant access to a resource.
+  MFA_REQUIRED_YES = 1;
+  // Completing an MFA ceremony will not grant access to a resource.
+  MFA_REQUIRED_NO = 2;
++ // Completing an MFA ceremony or reusing the MFA response of the required
++ // scope will grant access to a resource.
++ MFA_REQUIRED_YES_ALLOW_REUSE = 3;
```

Comment:
Is this added so the client can determine whether it needs a new challenge?

Isn't reuse tangential to MFARequired yes/no? Should we add a new type to capture reuse yes/no instead?